File: frontend/nextjs-app/src/lib/services/quran-service.ts
/**
 * Quran service for API interactions
 */
import { apiClient } from '../api/axios-client';
import { API_ENDPOINTS } from '../api/endpoints';
import type { Surah, Juz, QuranBookmark, ReadingProgress, QuranPage, Ayah } from '@/types/quran';

export class QuranService {
  /**
   * Get all surahs
   */
  static async getSurahs(): Promise<Surah[]> {
    try {
      return await apiClient.get<Surah[]>(API_ENDPOINTS.QURAN.SURAHS);
    } catch (error) {
      console.error('Failed to fetch surahs:', error);
      throw new Error('Failed to load surahs');
    }
  }

  /**
   * Get a specific surah by number
   */
  static async getSurah(surahNumber: number): Promise<Surah> {
    try {
      return await apiClient.get<Surah>(
        API_ENDPOINTS.QURAN.SURAH(surahNumber)
      );
    } catch (error) {
      console.error(`Failed to fetch surah ${surahNumber}:`, error);
      throw new Error(`Failed to load surah ${surahNumber}`);
    }
  }

  /**
   * Get all juzs
   */
  static async getJuzs(): Promise<Juz[]> {
    try {
      return await apiClient.get<Juz[]>(API_ENDPOINTS.QURAN.JUZS);
    } catch (error) {
      console.error('Failed to fetch juzs:', error);
      throw new Error('Failed to load juzs');
    }
  }

  /**
   * Get a specific juz by number
   */
  static async getJuz(juzNumber: number): Promise<Juz> {
    try {
      return await apiClient.get<Juz>(
        API_ENDPOINTS.QURAN.JUZ(juzNumber)
      );
    } catch (error) {
      console.error(`Failed to fetch juz ${juzNumber}:`, error);
      throw new Error(`Failed to load juz ${juzNumber}`);
    }
  }

  /**
   * Search surahs by name or number
   */
  static async searchSurahs(query: string): Promise<Surah[]> {
    try {
      return await apiClient.get<Surah[]>(API_ENDPOINTS.QURAN.SURAHS, {
        params: { search: query },
      });
    } catch (error) {
      console.error('Failed to search surahs:', error);
      throw new Error('Failed to search surahs');
    }
  }

  /**
   * Get user bookmarks
   */
  static async getBookmarks(): Promise<QuranBookmark[]> {
    try {
      return await apiClient.get<QuranBookmark[]>(
        API_ENDPOINTS.USER_BOOKMARKS
      );
    } catch (error) {
      console.error('Failed to fetch bookmarks:', error);
      throw new Error('Failed to load bookmarks');
    }
  }

  /**
   * Add a bookmark
   */
  static async addBookmark(data: {
    surah_number: number;
    ayah_number: number;
    page_number: number;
    note?: string;
  }): Promise<QuranBookmark> {
    try {
      return await apiClient.post<QuranBookmark>(
        API_ENDPOINTS.USER_BOOKMARKS,
        data
      );
    } catch (error) {
      console.error('Failed to add bookmark:', error);
      throw new Error('Failed to add bookmark');
    }
  }

  /**
   * Delete a bookmark
   */
  static async deleteBookmark(bookmarkId: string): Promise<void> {
    try {
      await apiClient.delete(`${API_ENDPOINTS.USER_BOOKMARKS}/${bookmarkId}`);
    } catch (error) {
      console.error('Failed to delete bookmark:', error);
      throw new Error('Failed to delete bookmark');
    }
  }

  /**
   * Get reading progress
   */
  static async getReadingProgress(): Promise<ReadingProgress> {
    try {
      return await apiClient.get<ReadingProgress>(
        API_ENDPOINTS.USER_READING_PROGRESS
      );
    } catch (error) {
      console.error('Failed to fetch reading progress:', error);
      throw new Error('Failed to load reading progress');
    }
  }

  /**
   * Update reading progress
   */
  static async updateReadingProgress(data: {
    surah_number: number;
    ayah_number: number;
    page_number: number;
  }): Promise<void> {
    try {
      await apiClient.post(API_ENDPOINTS.USER_READING_PROGRESS, data);
    } catch (error) {
      console.error('Failed to update reading progress:', error);
      throw new Error('Failed to update reading progress');
    }
  }

  /**
   * Get a specific page of the Quran
   */
  static async getPage(pageNumber: number): Promise<QuranPage> {
    try {
      return await apiClient.get<QuranPage>(`/api/quran/pages/${pageNumber}`);
    } catch (error) {
      console.error(`Failed to fetch page ${pageNumber}:`, error);
      throw new Error(`Failed to load page ${pageNumber}`);
    }
  }

  /**
   * Get ayahs for a specific surah
   */
  static async getSurahAyahs(surahNumber: number): Promise<Ayah[]> {
    try {
      return await apiClient.get<Ayah[]>(`/api/quran/surahs/${surahNumber}/ayahs`);
    } catch (error) {
      console.error(`Failed to fetch ayahs for surah ${surahNumber}:`, error);
      throw new Error(`Failed to load ayahs for surah ${surahNumber}`);
    }
  }

  /**
   * Get ayahs for a specific page
   */
  static async getPageAyahs(pageNumber: number): Promise<Ayah[]> {
    try {
      return await apiClient.get<Ayah[]>(`/api/quran/pages/${pageNumber}/ayahs`);
    } catch (error) {
      console.error(`Failed to fetch ayahs for page ${pageNumber}:`, error);
      throw new Error(`Failed to load ayahs for page ${pageNumber}`);
    }
  }
}

File: frontend/mobile/lib/core/services/quran_service.dart
import 'package:dio/dio.dart';
import '../network/dio_client.dart';
import '../network/api_endpoints.dart';
import '../../features/quran/data/models/surah_model.dart';
import '../../features/quran/data/models/ayah_model.dart';

/// Service for Quran-related API calls
class QuranService {
  final DioClient _dioClient;

  QuranService(this._dioClient);

  /// Get all surahs
  Future<List<SurahModel>> getSurahs() async {
    try {
      final response = await _dioClient.get(ApiEndpoints.surahs);
      final List<dynamic> data = response.data as List<dynamic>;
      return data.map((json) => SurahModel.fromJson(json as Map<String, dynamic>)).toList();
    } catch (e) {
      throw Exception('Failed to load surahs: $e');
    }
  }

  /// Get a specific surah by number
  Future<SurahModel> getSurah(int surahNumber) async {
    try {
      final response = await _dioClient.get(ApiEndpoints.surah(surahNumber));
      return SurahModel.fromJson(response.data as Map<String, dynamic>);
    } catch (e) {
      throw Exception('Failed to load surah $surahNumber: $e');
    }
  }

  /// Get all juzs
  Future<List<JuzModel>> getJuzs() async {
    try {
      final response = await _dioClient.get(ApiEndpoints.juzs);
      final List<dynamic> data = response.data as List<dynamic>;
      return data.map((json) => JuzModel.fromJson(json as Map<String, dynamic>)).toList();
    } catch (e) {
      throw Exception('Failed to load juzs: $e');
    }
  }

  /// Get a specific juz by number
  Future<JuzModel> getJuz(int juzNumber) async {
    try {
      final response = await _dioClient.get(ApiEndpoints.juz(juzNumber));
      return JuzModel.fromJson(response.data as Map<String, dynamic>);
    } catch (e) {
      throw Exception('Failed to load juz $juzNumber: $e');
    }
  }

  /// Search surahs by name or number
  Future<List<SurahModel>> searchSurahs(String query) async {
    try {
      final response = await _dioClient.get(
        ApiEndpoints.surahs,
        queryParameters: {'search': query},
      );
      final List<dynamic> data = response.data as List<dynamic>;
      return data.map((json) => SurahModel.fromJson(json as Map<String, dynamic>)).toList();
    } catch (e) {
      throw Exception('Failed to search surahs: $e');
    }
  }

  /// Get user bookmarks
  Future<List<QuranBookmark>> getBookmarks() async {
    try {
      final response = await _dioClient.get(ApiEndpoints.userBookmarks);
      final List<dynamic> data = response.data as List<dynamic>;
      return data.map((json) => QuranBookmark.fromJson(json as Map<String, dynamic>)).toList();
    } catch (e) {
      throw Exception('Failed to load bookmarks: $e');
    }
  }

  /// Add a bookmark
  Future<QuranBookmark> addBookmark({
    required int surahNumber,
    required int ayahNumber,
    required int pageNumber,
    String? note,
  }) async {
    try {
      final response = await _dioClient.post(
        ApiEndpoints.userBookmarks,
        data: {
          'surah_number': surahNumber,
          'ayah_number': ayahNumber,
          'page_number': pageNumber,
          'note': note,
        },
      );
      return QuranBookmark.fromJson(response.data as Map<String, dynamic>);
    } catch (e) {
      throw Exception('Failed to add bookmark: $e');
    }
  }

  /// Delete a bookmark
  Future<void> deleteBookmark(String bookmarkId) async {
    try {
      await _dioClient.delete('${ApiEndpoints.userBookmarks}/$bookmarkId');
    } catch (e) {
      throw Exception('Failed to delete bookmark: $e');
    }
  }

  /// Get reading progress
  Future<Map<String, dynamic>> getReadingProgress() async {
    try {
      final response = await _dioClient.get(ApiEndpoints.userReadingProgress);
      return response.data as Map<String, dynamic>;
    } catch (e) {
      throw Exception('Failed to load reading progress: $e');
    }
  }

  /// Update reading progress
  Future<void> updateReadingProgress({
    required int surahNumber,
    required int ayahNumber,
    required int pageNumber,
  }) async {
    try {
      await _dioClient.post(
        ApiEndpoints.userReadingProgress,
        data: {
          'surah_number': surahNumber,
          'ayah_number': ayahNumber,
          'page_number': pageNumber,
        },
      );
    } catch (e) {
      throw Exception('Failed to update reading progress: $e');
    }
  }

  /// Get a specific page of the Quran
  Future<QuranPageModel> getPage(int pageNumber) async {
    try {
      final response = await _dioClient.get('/api/quran/pages/$pageNumber');
      return QuranPageModel.fromJson(response.data as Map<String, dynamic>);
    } catch (e) {
      throw Exception('Failed to load page $pageNumber: $e');
    }
  }

  /// Get ayahs for a specific surah
  Future<List<AyahModel>> getSurahAyahs(int surahNumber) async {
    try {
      final response = await _dioClient.get('/api/quran/surahs/$surahNumber/ayahs');
      final List<dynamic> data = response.data as List<dynamic>;
      return data.map((json) => AyahModel.fromJson(json as Map<String, dynamic>)).toList();
    } catch (e) {
      throw Exception('Failed to load ayahs for surah $surahNumber: $e');
    }
  }

  /// Get ayahs for a specific page
  Future<List<AyahModel>> getPageAyahs(int pageNumber) async {
    try {
      final response = await _dioClient.get('/api/quran/pages/$pageNumber/ayahs');
      final List<dynamic> data = response.data as List<dynamic>;
      return data.map((json) => AyahModel.fromJson(json as Map<String, dynamic>)).toList();
    } catch (e) {
      throw Exception('Failed to load ayahs for page $pageNumber: $e');
    }
  }
}
